domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,354 @@
+//
+// It is a hardcoded version of architecture table in controller schema.
+var (
+	//
+	architectureIDsByName = map[string]int{
+		"amd64":   0,
+		"arm64":   1,
+		"ppc64el": 2,
+		"s390x":   3,
+		"riscv64": 4,
+	}
+)
+
+// State encapsulates the state management, logging, and architecture data.
+type State struct {
+	*domain.StateBase
+	clock  clock.Clock
+	logger logger.Logger
+}
+
+// NewState creates a new State instance using the provided database transaction factory and logger.
+func NewState(factory database.TxnRunnerFactory, clock clock.Clock, logger logger.Logger) *State {
+	state := &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+		clock:     clock,
+	}
+	return state
+}
+
+// getArchitectureID returns the ID for a given architecture name and a boolean indicating if the name was found.
+func getArchitectureID(name string) (int, bool) {
+	id, ok := architectureIDsByName[name]
+	return id, ok
+}
+
+// SaveMetadata stores the provided list of cloud image metadata into the database.
+//
+// Returns [cloudmetadataerrors.Invalid] if the provided metadata omits required fields.
+//
+// It also fires a cleanup for old images, if any.
+//
+// [cloudimagemetadata.Metadata] are considered unique among a signature, composed of these fields:
+//
+//   - Stream
+//   - Region
+//   - Version
+//   - Arch
+//   - VirtType
+//   - RootStorageType
+//   - Source
+//
+// Above behaviors applies for duplicated inserted metadata:
+//   - If a metadata has the same signature than an existing one in the database, the imageID will be
+//     updated with the new value in the existing metadata
+//   - If a several metadatas have the same signature in the list, it will cause a unique constraint error from
+//     the database. It is likely a programmatic error.
+func (s *State) SaveMetadata(ctx context.Context, metadatas []cloudimagemetadata.Metadata) error {
+	// Cleaning the database before insert/update is a reasonable way to remove outdated data without
+	// requiring an housekeeping goroutine. Since a metadata data doesn't live long in the db and
+	// it shouldn't be a lot of custom metadata, it shouldn't slow too much the insertion.
+	// Expired metadata are ignored when retrieving metadata, so it doesn't have any impact to keep
+	// some of them longer than their time to live.
+	if err := s.tryCleanUpExpiredMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare inputs
+	createdAt := s.clock.Now()
+	values := make([]inputMetadata, 0, len(metadatas))
+	for _, m := range metadatas {
+		// Convert architecture name to a db id
+		archId, ok := getArchitectureID(m.Arch)
+		if !ok {
+			s.logger.Warningf("skipping metadata for unknown architecture %q", m.Arch)
+			continue
+		}
+
+		// Update creation time if not defined.
+		var noTime time.Time
+		if m.CreationTime != noTime {
+			createdAt = m.CreationTime
+		}
```

Comment:
```suggestion
		if m.CreationTime.IsZero() {
			createdAt = m.CreationTime
		}
```

https://pkg.go.dev/time#Time.IsZero